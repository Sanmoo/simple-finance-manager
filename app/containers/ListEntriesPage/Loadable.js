/**
 *
 * Asynchronously loads the component for ListEntriesPage
 *
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
