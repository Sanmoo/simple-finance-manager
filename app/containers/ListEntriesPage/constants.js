/*
 *
 * ListEntriesPage constants
 *
 */

export const DEFAULT_ACTION = 'app/ListEntriesPage/DEFAULT_ACTION';
export const SAVE_KEY = 'app/ListEntriesPage/SAVE_KEY';
