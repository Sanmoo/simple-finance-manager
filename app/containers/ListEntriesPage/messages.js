/*
 * ListEntriesPage Messages
 *
 * This contains all the text for the ListEntriesPage container.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.containers.ListEntriesPage';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the ListEntriesPage container!',
  },
});
