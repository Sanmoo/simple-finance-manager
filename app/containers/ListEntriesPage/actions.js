/*
 *
 * ListEntriesPage actions
 *
 */

import { DEFAULT_ACTION, SAVE_KEY } from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function saveKey(key, value) {
  return {
    type: SAVE_KEY,
    key,
    value,
  };
}
