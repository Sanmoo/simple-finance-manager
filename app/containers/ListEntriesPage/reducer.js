/*
 *
 * ListEntriesPage reducer
 *
 */
import produce from 'immer';
import { DEFAULT_ACTION, SAVE_KEY } from './constants';

export const initialState = {
  entries: [],
};

/* eslint-disable default-case, no-param-reassign */
const listEntriesPageReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case DEFAULT_ACTION:
        break;
      case SAVE_KEY:
        draft[action.key] = action.value;
        break;
    }
  });

export default listEntriesPageReducer;
