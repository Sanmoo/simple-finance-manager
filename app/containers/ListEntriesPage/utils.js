import { sortBy } from 'lodash';
import { SPREADSHEET_DATE_FORMATS } from 'utils/constants';
import parse from 'date-fns/parse';

export const orderByDateDesc = entries =>
  sortBy(
    entries,
    e => parse(e.date, SPREADSHEET_DATE_FORMATS['pt-BR'], new Date()),
    'desc',
  );
