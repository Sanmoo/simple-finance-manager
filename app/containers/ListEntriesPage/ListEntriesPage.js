/**
 *
 * ListEntriesPage
 *
 */

import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Helmet } from 'react-helmet';
import { Card, CardContent } from '@material-ui/core';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import Paper from '@material-ui/core/Paper';
import TableRow from '@material-ui/core/TableRow';
import { withStyles } from '@material-ui/core/styles';
import {
  getSheetTitleForCurrentMonth,
  generateUniqueKeyForEntry,
  getAllEntriesForCurrentMonth,
} from 'utils/business';
import { SPREADSHEET_DATE_FORMATS } from 'utils/constants';
import { TYPE_EXPENSE, TYPE_INCOME } from 'utils/businessConstants';
import parse from 'date-fns/parse';
import format from 'date-fns/format';
import TableToolbar from './TableToolbar';

const extractDateDay = dateAsStr => {
  const date = parse(dateAsStr, SPREADSHEET_DATE_FORMATS['pt-BR'], new Date());
  return format(date, 'd');
};

const styles = () => ({
  root: {
    overflowX: 'auto',
  },
  cell: {
    paddingRight: 0,
  },
  cellValue: {
    minWidth: 100,
  },
  table: {
    minWidth: 400,
  },
});

const SHOWN_TYPE_TITLES = {
  [TYPE_EXPENSE]: 'Expenses',
  [TYPE_INCOME]: 'Income',
};

function ListEntriesPage({
  entries,
  classes,
  onViewExpensesClick,
  onViewIncomesClick,
  shownType,
  onEntriesLoaded,
  onEditEntryClick,
}) {
  useEffect(() => {
    getAllEntriesForCurrentMonth().then(onEntriesLoaded);
  }, []);

  return (
    <div>
      <Helmet>
        <title>Entries List</title>
        <meta
          name="description"
          content="Page for listing the entries you already provided"
        />
      </Helmet>
      <Card className="content">
        <CardContent>
          <Paper className={classes.root}>
            <TableToolbar
              title={`${
                SHOWN_TYPE_TITLES[shownType]
              } ${getSheetTitleForCurrentMonth()}`}
              {...{ onViewExpensesClick, onViewIncomesClick }}
            />
            <Table className={classes.table}>
              <TableBody>
                {entries.map(entry => (
                  <TableRow
                    key={generateUniqueKeyForEntry(entry)}
                    onClick={() => onEditEntryClick(entry)}
                  >
                    <TableCell className={classes.cell}>
                      {extractDateDay(entry.date)}
                    </TableCell>
                    <TableCell align="left" className={classes.cell}>
                      {entry.desc}
                    </TableCell>
                    <TableCell align="left" className={classes.cell}>
                      {entry.category}
                    </TableCell>
                    <TableCell
                      align="left"
                      className={classNames(classes.cell, classes.cellValue)}
                    >
                      {entry.value}
                    </TableCell>
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </Paper>
        </CardContent>
      </Card>
    </div>
  );
}

ListEntriesPage.propTypes = {
  classes: PropTypes.object.isRequired,
  entries: PropTypes.arrayOf(
    PropTypes.shape({
      date: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      originSheetTitle: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
      category: PropTypes.string.isRequired,
      desc: PropTypes.string.isRequired,
    }),
  ).isRequired,
  location: PropTypes.shape({
    search: PropTypes.string,
  }),
  onEntriesLoaded: PropTypes.func.isRequired,
  onViewExpensesClick: PropTypes.func.isRequired,
  shownType: PropTypes.oneOf([TYPE_EXPENSE, TYPE_INCOME]),
  onViewIncomesClick: PropTypes.func.isRequired,
  onEditEntryClick: PropTypes.func.isRequired,
};

export default withStyles(styles)(ListEntriesPage);
