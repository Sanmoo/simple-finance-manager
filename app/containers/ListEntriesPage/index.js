/**
 *
 * ListEntriesPage
 *
 */

import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import { makeSelectEntries } from './selectors';
import reducer from './reducer';
import saga from './saga';
import ListEntriesPage from './ListEntriesPage';
import { saveKey } from './actions';

const mapStateToProps = createStructuredSelector({
  entries: makeSelectEntries(),
});

function mapDispatchToProps(dispatch) {
  return {
    onEntriesLoaded: entries => dispatch(saveKey('entries', entries)),
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withReducer = injectReducer({ key: 'listEntriesPage', reducer });
const withSaga = injectSaga({ key: 'listEntriesPage', saga });

export default compose(
  withConnect,
  withReducer,
  withSaga,
)(ListEntriesPage);
