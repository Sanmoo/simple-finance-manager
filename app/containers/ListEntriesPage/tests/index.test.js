/**
 *
 * Tests for ListEntriesPage
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { TYPE_EXPENSE } from 'utils/businessConstants';
import 'fake-indexeddb/auto';

import ListEntriesPage from '../ListEntriesPage';
import { DEFAULT_LOCALE } from '../../../i18n';

describe('<ListEntriesPage />', () => {
  let props;

  beforeEach(() => {
    props = {
      entries: [
        {
          date: '10/11/2019',
          type: 'type',
          originSheetTitle: 'May-2019',
          value: 20,
          category: 'Category A',
          desc: 'My Desc',
        },
      ],
      location: {
        search: 'my search',
      },
      onEntriesLoaded: jest.fn(),
      shownType: TYPE_EXPENSE,
      onViewIncomesClick: jest.fn(),
      onViewExpensesClick: jest.fn(),
      onEditEntryClick: jest.fn(),
    };
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ListEntriesPage {...props} />
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ListEntriesPage {...props} />
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
