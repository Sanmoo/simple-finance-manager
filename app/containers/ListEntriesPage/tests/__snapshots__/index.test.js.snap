// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListEntriesPage /> Should render and match the snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root-140 MuiPaper-elevation1-143 MuiPaper-rounded-141 MuiCard-root-139 content"
  >
    <div
      class="MuiCardContent-root-167"
    >
      <div
        class="MuiPaper-root-140 MuiPaper-elevation2-144 MuiPaper-rounded-141 ListEntriesPage-root-135"
      >
        <div
          class="MuiToolbar-root-173 MuiToolbar-regular-175 MuiToolbar-gutters-174 TableToolbar-root-168"
        >
          <div
            class="TableToolbar-title-172"
          >
            <h6
              class="MuiTypography-root-177 MuiTypography-h6-194"
              id="tableTitle"
            >
              Expenses Jun-2019
            </h6>
          </div>
          <div
            class="TableToolbar-spacer-170"
          />
          <div
            class="TableToolbar-actions-171"
          >
            <button
              aria-label="Filters"
              class="MuiButtonBase-root-227 MuiIconButton-root-221"
              tabindex="0"
              title="Filters"
              type="button"
            >
              <span
                class="MuiIconButton-label-226"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root-230"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root-262"
              />
            </button>
          </div>
        </div>
        <table
          class="MuiTable-root-243 ListEntriesPage-table-138"
        >
          <tbody
            class="MuiTableBody-root-244"
          >
            <tr
              class="MuiTableRow-root-245"
            >
              <td
                class="MuiTableCell-root-250 MuiTableCell-body-252 ListEntriesPage-cell-136"
              >
                10
              </td>
              <td
                class="MuiTableCell-root-250 MuiTableCell-body-252 MuiTableCell-alignLeft-258 ListEntriesPage-cell-136"
              >
                My Desc
              </td>
              <td
                class="MuiTableCell-root-250 MuiTableCell-body-252 MuiTableCell-alignLeft-258 ListEntriesPage-cell-136"
              >
                Category A
              </td>
              <td
                class="MuiTableCell-root-250 MuiTableCell-body-252 MuiTableCell-alignLeft-258 ListEntriesPage-cell-136 ListEntriesPage-cellValue-137"
              >
                20
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
